proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,138 @@
+# Feature name
+
+* Proposal: TBD
+* Author(s): [Kevin Ballard](https://github.com/kballard)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+Add 3 new `SequenceType` functions `scan(_:combine:)`, `takeWhile(_:)`, and
+`dropWhile(_:)`, with overrides as appropriate on `CollectionType`,
+`LazySequenceType`, and `LazyCollectionType`, as well as a global function
+`iterate(_:apply:)`.
+
+Swift-evolution thread:
+[Proposal: Add scan, takeWhile, dropWhile, and iterate to the stdlib](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20151228/004690.html)
+
+## Motivation
+
+The Swift standard library provides many useful sequence manipulators like
+`dropFirst(_:)`, `filter(_:)`, etc. but it's missing a few common methods that
+are quite useful.
+
+## Proposed solution
+
+Add the following extension to `SequenceType`:
+
+```swift
+extension SequenceType {
+  /// Returns an array containing the results of
+  ///
+  ///     p.reduce(initial, combine: combine)
+  ///
+  /// for each prefix `p` of `self`, in order from shortest to longest.
+  /// For example:
+  ///
+  ///     (1..<6).scan(0, combine: +) // [0, 1, 3, 6, 10, 15]
+  ///
+  /// - Complexity: O(N)
+  func scan<T>(initial: T, @noescape combine: (T, Self.Generator.Element) throws -> T) rethrows -> [T]
+}
+```
+
+Modify the declaration of `SequenceType` with two new members:
+
+```swift
+protocol SequenceType {
+  // ...
+  /// Returns a subsequence by skipping elements while `dropElement` returns
+  /// `true` and returning the remainder.
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  /// Returns a subsequence containing the elements until `takeElement` returns
+  /// `false` and skipping the remainder.
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+}
+```
+
+Also provide default implementations on `SequenceType` that return
+`AnySequence`, and default implementations on `CollectionType` that return a
+slice.
+
+`LazySequenceType` and `LazyCollectionType` will also be extended with
+implementations of `scan(_:combine:)`, `dropWhile(_:)`, and `takeWhile(_:)`
+that return lazy sequence/collection types. Like the lazy `filter(_:)`,
+`dropWhile(_:)` will perform the filtering when `startIndex` is accessed.
+
+Add a global function:
+
+```swift
+/// Returns an infinite sequence of lazy applications of `apply` to the
+/// previous value. For example:
+///
+///     iterate(1, apply: { $0 * 2 }) // yields: 1, 2, 4, 8, 16, 32, 64, ...
+func iterate<T>(initial: T, apply: T -> T) -> IterateSequence<T>
```

Comment:
`scan(_:combine:)` returns an array, you can't use it on an infinite sequence. The lazy version returns a lazy sequence, but the stdlib philosophy (last I checked) was to not perform lazy sequence operations without the user specifying `.lazy` somewhere, so this theoretical `repeatForever()` would have to be like `zip()` in that it may operate lazily itself but it still exposes eager sequence operations. So this turns the expression into `repeatForever(f).lazy.scan(0, apply: { $0 * 2 })`, which is a bit of a mouthful and isn't really the most obvious way to accomplish this. And I'm sure if I were to propose a `repeatForever(_:)` there'd be pushback from people saying that we really should just have a `.cycle()` method (which I suggested on the ML before and had a lot of debate over infinite sequences) instead, which would turn it onto the even more unwieldy `CollectionOfOne(f).cycle().lazy.scan(0, apply: { $0 * 2 })`.

All that said, I do want to introduce a `.cycle()` method, but `iterate(_:apply:)` is a specific pattern that is relatively common to encounter when doing FP-style manipulation of sequences (and works very nicely with `.takeWhile(_:)`) that it seems worth having, if for no other reason than the fact that it would save everyone who wants this type of functionality from having to redefine it themselves in terms of `CollectionOfOne(f).cycle().lazy.scan(_:apply:)`.